File: services/serviceData/PogList.py
"""
GPL 3 file header
"""
import copy
import uuid

from services.serviceData.PogData import PogData


class PogList:

	def __init__(self):
		self.pogList = []
		self.listVersion = 0

	def construct(self):
		dl = PogList()
		self.cloneData(dl)
		return dl

	def clone(self):
		pl = PogList()
		pl.listVersion = self.listVersion
		for pog in self.pogList:
			np = copy.copy(pog)
			np.uuid = str(uuid.uuid4())
			pl.pogList.append(np)
		return pl

	def cloneData(self, dl):
		if 'listVersion' in locals():
			dl.listVersion = self.listVersion
		else:
			dl.listVersion = 0
		if self.pogList is None:
			return
		for pog in self.pogList:
			pd = PogData.load(pog)
			pd1 = PogData()
			pd1.fullUpdate(pd)
			pd1.uuid = pd.uuid
			dl.pogList.append(pd1)

	def getListVersion(self):
		return self.listVersion

	def remove(self, pog):
		oldList = self.pogList
		self.pogList = []
		for pogInList in oldList:
			if not pogInList.isEqual(pog):
				self.pogList.append(pogInList)
		self.listVersion = self.listVersion + 1

	def addPog(self, pogToAdd):
		self.pogList.append(pogToAdd)
		self.listVersion += 1

	def update(self, src, dst):
		dst.fullUpdate(src)
		self.listVersion = self.listVersion + 1

	def addOrUpdate(self, pog):
		for pogInList in self.pogList:
			if pogInList.isEqual(pog):
				pogInList.fullUpdate(pog)
				self.listVersion += 1
				return
		self.addPog(pog)

File: linuxDist/BattleMatServer/webApp/WEB-INF/classes/per/lambert/ebattleMat/client/services/serviceData/PogList.java
/*
 * Copyright (C) 2019 Leon Lambert.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package per.lambert.ebattleMat.client.services.serviceData;

import com.google.gwt.core.client.JavaScriptObject;

/**
 * List of pogs.
 * 
 * @author LLambert
 *
 */
public class PogList extends JavaScriptObject {
	/**
	 * Constructor.
	 */
	protected PogList() {
	}

	/**
	 * List of pogs.
	 * 
	 * @return list of pogs.
	 */
	public final native PogData[] getPogList() /*-{
		if (this.pogList === undefined) {
			this.pogList = [];
		}
		return this.pogList;
	}-*/;

	public final void addPog(final PogData pogToAdd) {
		addPogNative(pogToAdd);
		setListVersion(getListVersion() + 1);
	}
	/**
	 * Add pog to list.
	 * 
	 * @param pogToAdd pog to add
	 */
	public final native void addPogNative(PogData pogToAdd) /*-{
		if (this.pogList === undefined) {
			this.pogList = [];
		}
		this.pogList.push(pogToAdd);
	}-*/;

	/**
	 * remove pog.
	 * 
	 * @param pog to remove
	 */
	public final void remove(final PogData pog) {
		PogData[] list = getPogList();
		initList();
		for (PogData pogInList : list) {
			if (!pogInList.isEqual(pog)) {
				addPogNative(pogInList);
			}
		}
		setListVersion(getListVersion() + 1);
	}

	public final void update(final PogData src, final PogData dst) {
		dst.fullUpdate(src);
		setListVersion(getListVersion() + 1);
	}

	/**
	 * create new list.
	 */
	private native void initList() /*-{
		this.pogList = [];
	}-*/;

	/**
	 * Get version of list.
	 * @return version of list
	 */
	public final native int getListVersion() /*-{
		if (this.listVersion === undefined) {
			this.listVersion = 0;
		}
		return this.listVersion;
	}-*/;

	/**
	 * set version of list.
	 * @param listVersion
	 */
	private native void setListVersion(int listVersion) /*-{
		this.listVersion = listVersion;
	}-*/;
}
